extensions/windows-event-log/CollectorInitiatedSubscription.cpp

```diff
@@ -0,0 +1,789 @@
+    {EcSubscriptionDeliveryMode, (uint32_t)EcDeliveryModePull},
+    {EcSubscriptionDeliveryMaxItems, (uint32_t)s_maxDeliveryItems.value()},
+    {EcSubscriptionDeliveryMaxLatencyTime, (uint32_t)s_deliveryMaxLatencyTime.value()},
+    {EcSubscriptionHeartbeatInterval, (uint32_t)s_heartbeatInterval.value()},
+    {EcSubscriptionContentFormat, (uint32_t)EcContentFormatRenderedText},
+    {EcSubscriptionCredentialsType, (uint32_t)EcSubscriptionCredDefault},
+    {EcSubscriptionEnabled, true},
+    {EcSubscriptionCommonUserName, s_sourceUserName.value()},
+    {EcSubscriptionCommonPassword, s_sourcePassword.value()}
+  };
+  for (auto prop : vProp) {
+    if (!EcSetSubscriptionProperty(hSubscription, prop.propId_, 0, &prop.prop_)) {
+      LOG_SUBSCRIPTION_WINDOWS_ERROR("EcSetSubscriptionProperty id: " + std::to_string(prop.propId_));
+      return false;
+    }
+  }
+
+  // Get the EventSources array so a new event source can be added for the specified target.
+  std::vector<BYTE> buffer;
+  PEC_VARIANT vProperty = NULL;
+  if (!getSubscriptionProperty(hSubscription, EcSubscriptionEventSources, 0, buffer, vProperty))
+    return false;
+
+  // Event Sources is a collection. Ensure that we have obtained handle to the Array Property. 
+  if (vProperty->Type != EcVarTypeNull && vProperty->Type != EcVarObjectArrayPropertyHandle) {
+    logInvalidSubscriptionPropertyType(__LINE__, vProperty->Type);
+    return false;
+  }
+
+  if (vProperty->Type == EcVarTypeNull) {
+    LOG_SUBSCRIPTION_ERROR("!hArray");
+    return false;
+  }
+
+  EC_OBJECT_ARRAY_PROPERTY_HANDLE hArray = vProperty->PropertyHandleVal;
+  utils::ScopeGuard guard_hArray([hArray]() { EcClose(hArray); });
+
+  DWORD dwEventSourceCount{};
+  if (!EcGetObjectArraySize(hArray, &dwEventSourceCount)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetObjectArraySize");
+    return false;
+  }
+
+  // Add a new EventSource to the EventSources array object.
+  if (!EcInsertObjectArrayElement(hArray, dwEventSourceCount)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcInsertObjectArrayElement");
+    return false;
+  }
+
+  for (auto& prop: std::vector<SubscriptionProperty>{{EcSubscriptionEventSourceAddress, s_sourceAddress.value()}, {EcSubscriptionEventSourceEnabled, true}}) {
+    if (!EcSetObjectArrayProperty(hArray, prop.propId_, dwEventSourceCount, 0, &prop.prop_)) {
+      LOG_SUBSCRIPTION_WINDOWS_ERROR("EcSetObjectArrayProperty id: " + std::to_string(prop.propId_));
+      return false;
+    }
+  }
+
+  if (!EcSaveSubscription(hSubscription, NULL)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcSaveSubscription");
+    return false;
+  }
+
+  return true;
+}
+
+bool CollectorInitiatedSubscription::subscribe(const std::shared_ptr<core::ProcessContext> &context)
+{
+  logger_->log_debug("CollectorInitiatedSubscription: maxBufferSize_ %lld", s_maxBufferSize.value());
+
+  provenanceUri_ = "winlog://" + computerName_ + "/" + to_string(s_channel.value()) + "?" + to_string(s_query.value());
```

Comment:
Although this is not a big deal, I would use std::string::append here.

Reply:
Since it is not a big deal, will leave it as it is :-)